### DISCUSSION TITLE
Unauthenticated client/bot prevention

### DISCUSSION THREAD

**Discussion by pplulee** [postalserver/postal#2791] [Feature suggestions]
Recently my logs have been showing a large number of machine scripts accessing the SMTP port that don't authenticate and try to send mail, taking up a lot of resources.

I've tried using fail2ban for protection(#1255), but the situation is different to failed authentication attempts(#1182), currently the log context needs to be analysed to get the attacker's IP address.

Is there any solution to prevent this kind of attack, or is it possible to improve the error log display to fit the simple regex?

![image](https://github.com/postalserver/postal/assets/46434255/f7ef5c30-b13c-45c1-a10d-219d6b5be48a)

**Comment by Pacerino** [comment]
This is the solution with fail2ban i'm using to prevent bot spamming looking for open relays.
https://github.com/postalserver/postal/pull/1187#issuecomment-683939769

**Comment by pplulee** [comment]
Thanks for the help, but things are a little different here. The attack pattern I met is bot only sends lots of 'RCPT TO' commands before authentication, and then disconnects. The IP of the attacker client can only be found when it establishes the connection. Fail2ban cannot fit this log pattern(although I wrote a script to monitor the log and use fail2ban to ban IPs).
![image](https://github.com/postalserver/postal/assets/46434255/9a1546a3-bc5d-4277-a678-e24b8f5646a1)
I wonder if it is possible to improve the format of 'Authentication error'?

**Reply by Pacerino** [reply]
I see multiple approaches here:

1. Add a log entry with the ip address when the `530 Authentication required` will be send https://github.com/postalserver/postal/blob/main/lib/postal/smtp_server/client.rb#L353
2. Add the client ip to each log entry https://github.com/postalserver/postal/blob/main/lib/postal/smtp_server/client.rb#L101
3. Finally: this procedure would not require a change to the code. Each connection/log entry is provided with a unique ID. What you could do would be to use the ID to get the IP address and then count the error message `530 Authentication required` per ID. If this value is exceeded, the IP address belonging to this ID is simply blocked.

**Reply by Pacerino** [reply]
Adding `log "\e[33m   WARN: AUTH failure for #{@ip_address}\e[0m"` above this line https://github.com/postalserver/postal/blob/main/lib/postal/smtp_server/client.rb#L353 would show an warning with the client ip address. The same line is used in failed authentication, therefore the fail2ban jail should also work for that.

**Reply by pplulee** [reply]
This is an effective workaround. After I added the output error message, fail2ban can easily identify the log and ban the IP.
![image](https://github.com/postalserver/postal/assets/46434255/f4cc617a-def1-47f9-a02d-71a8f0ba5a65)
